src/crypto.js

```diff
@@ -96,47 +112,76 @@ const hash256 = async (msg, type = 'SHA-256') => {
   const _salt = salt || genRandomBuffer(16)
   const iterations = 100000
-  const hashedValue = await deriveBitsAndHash(passPhrase, _salt, iterations)
+  const encMK = await deriveBitsGenAndEncMK(passPhrase, _salt, iterations, hashAlgo)
   return {
     salt: Buffer.from(_salt).toString('hex'),
     iterations: iterations,
-    hashAlgo: 'SHA-256',
-    storedHash: Buffer.from(hashedValue).toString('hex')
+    hashAlgo,
+    encMK
   }
 }
 
 /**
- * Derive the passphrase with PBKDF2 and hash the output with the given hash function
+ * Derive the passphrase with PBKDF2
+ * Generate a AES key (MK)
+ * Encrypt the MK
  *
  * @param {string | arrayBuffer} passPhrase The passphrase that is used to derive the key
  * @param {arrayBuffer} [salt] The salt
  * @param {Number} [iterations] The iterations number
  * @param {string} [hash] The hash function used for derivation and final hash computing
  * @returns {Promise<Uint8Array>}   A promise that contains the hashed derived key
  */
-const deriveBitsAndHash = async (passPhrase, salt, iterations, hash) => {
+const deriveBitsGenAndEncMK = async (passPhrase, salt, iterations, hash) => {
   const derivedPassphrase = await deriveBits(passPhrase, salt, iterations, hash)
-  const finalHash = await hash256(derivedPassphrase)
-  return finalHash
+  const KEK = await importKey(derivedPassphrase)
+  const MK = genRandomBuffer(16)
+  const encMK = await encrypt(KEK, Buffer.from(MK).toString('hex'))
+  return encMK
+}
+
+/**
+ * Derive the passphrase with PBKDF2 to obtain the KEK
+ * Decrypt the encypted MK
+ * return the raw MK
+ *
+ * @param {string | arrayBuffer} passPhrase The passphrase that is used to derive the key
+ * @param {arrayBuffer} [salt] The salt
+ * @param {Number} [iterations] The iterations number
+ * @param {string} [hash] The hash function used for derivation and final hash computing
+ * @param {encMK} [encMK] The encrypted MK
+ * @returns {Promise<Uint8Array>}   A promise that contains the hashed derived key
+ */
+const deriveBitsDecMK = async (passPhrase, salt, iterations, hash, encMK) => {
+  const _salt = typeof (salt) === 'string' ? Buffer.from(salt, ('hex')) : salt
+  const derivedPassphrase = await deriveBits(passPhrase, _salt, iterations, hash)
+  const KEK = await importKey(derivedPassphrase)
+  return decrypt(KEK, encMK)
 }
 
-const requiredParameterHashedPassphrase = ['salt', 'iterations', 'storedHash', 'hashAlgo']
+const requiredParameterProtectedMK = ['salt', 'iterations', 'encMK', 'hashAlgo']
 
 /**
  * Check a given passphrase by comparing it to the stored hash value (in HashedPassphrase object)
  *
  * @param {string} passphrase The passphrase
- * @param {HashedPassphrase} hashedPassphrase The HashedPassphrase object
+ * @param {protectedMK} protectedMK The protectedMK object
  * @returns {Promise<Boolean>}   A promise
  */
-const checkPassphrase = async (passPhrase, hashedPassphrase) => {
+const checkPassphrase = async (passPhrase, protectedMK) => {
```

Comment:
we should rename this methods to something like `getMasterKey`, what do you think ?

Reply:
yes right I will check Sujeeban proposition.